config/environment/ui-test.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+return array(
+
+    'Piwik\Config' => \DI\decorate(function (\Piwik\Config $config) {
+        $config->General['cors_domains'][] = '*';
+        $config->trusted_hosts[] = $config->tests['http_host'];
+        $config->trusted_hosts[] = $config->tests['http_host'] . ':' . $config->tests['port'];
+        return $config;
+    }),
+
+    'observers.global' => \DI\add(array(
+
+        array('Request.dispatch.end', function (&$result) {
```

Comment:
Only if the event does not belong to a plugin or is dependent on the DI environment. So if you want to call `Piwik::addAction` before the DI container is created, you would have to use something like this. Or if you want the event to be registered only in one DI environment (ie, test, dev, ui-test, cli, whatever), you would have to do it this way. It's not normal though, something core should only do.